flang/lib/Lower/OpenMP/ClauseProcessor.cpp

```diff
@@ -619,9 +620,28 @@ bool ClauseProcessor::processCopyin() const {
               checkAndCopyHostAssociateVar(&*mem, &insPt);
             break;
           }
-          if (semantics::IsAllocatableOrObjectPointer(&sym->GetUltimate()))
-            TODO(converter.getCurrentLocation(),
-                 "pointer or allocatable variables in Copyin clause");
+          if (semantics::IsAllocatable(sym->GetUltimate())) {
+            // copyin should copy the association of the allocatable
+
+            fir::FirOpBuilder &builder = converter.getFirOpBuilder();
+            const Fortran::semantics::Symbol &hsym = sym->GetUltimate();
+
+            Fortran::lower::SymbolBox hsb =
+                converter.lookupOneLevelUpSymbol(hsym);
+            assert(hsb && "Host symbol box not found");
+
+            Fortran::lower::SymbolBox sb = converter.shallowLookupSymbol(*sym);
+            assert(sb && "Host-associated symbol box not found");
+            assert(hsb.getAddr() != sb.getAddr() &&
+                   "Host and associated symbol boxes are the same");
+
+            mlir::Location loc = converter.genLocation(sym->name());
+            mlir::OpBuilder::InsertionGuard ipGuard{builder};
+            builder.setInsertionPointAfter(sb.getAddr().getDefiningOp());
+            builder.create<hlfir::AssignOp>(loc, hsb.getAddr(), sb.getAddr(),
+                                            true, false, false);
+          }
```

Comment:
Can't `copyHostAssociateVar` be used instead?

After https://github.com/llvm/llvm-project/pull/106559, it should be easier to use it to this end and avoid code duplication.

Reply:
`copyHostAssociateVar` is already being called at line 647 but that wasn't sufficient in my testing to mean that the allocatable got copied in. I needed to add the `hlfir.assign` to get the correct behaviour. I was testing with the following program:

```fortran
program main
  integer, allocatable, save :: a
  !$omp threadprivate(a)
  
  !$omp master
  allocate(a)
  a = 1
  !$omp end master
  
  !$omp parallel copyin(a)
  if (allocated(a)) then
    print *, a
  endif
  !$omp end parallel
end program
```

Without the copyin(a) this should only print once, whereas with the copyin(a) it should print OMP_NUM_THREADS times. Without the code block highlighted here, it only prints once.

Reply:
The test program works correctly with https://github.com/llvm/llvm-project/pull/106559, if I just remove the TODO:
```
--- a/flang/lib/Lower/OpenMP/ClauseProcessor.cpp
+++ b/flang/lib/Lower/OpenMP/ClauseProcessor.cpp
@@ -619,9 +619,6 @@ bool ClauseProcessor::processCopyin() const {
               checkAndCopyHostAssociateVar(&*mem, &insPt);
             break;
           }
-          if (semantics::IsAllocatableOrObjectPointer(&sym->GetUltimate()))
-            TODO(converter.getCurrentLocation(),
-                 "pointer or allocatable variables in Copyin clause");
           assert(sym->has<semantics::HostAssocDetails>() &&
                  "No host-association found");
           checkAndCopyHostAssociateVar(sym);
```

But line https://github.com/llvm/llvm-project/pull/106559/files#diff-0f70165af00b05acb5abeb11198079b793a20acf8f069edd2dbe7b34ac67b328R1262 should be chaged too, to allow `copyin` with unallocated variables.

Reply:
Are you saying, if we remove the todo and also add `|| flags.test(...::CopyIn` in the above place things should work?
When I _just_ removed the TODO it didn't work for me.

Reply:
Ah, I didn't have the above patch though. I'll have a look at this tomorrow (it's getting late here)

Reply:
> Are you saying, if we remove the todo and also add `|| flags.test(...::CopyIn` in the above place things should work? When I _just_ removed the TODO it didn't work for me.

Yes. I have just merged the patch to main.

Reply:
You're right, it's sufficient to just remove the TODO with your patch. Thanks!